Detectors/ITSMFT/MFT/base/src/GeometryTGeo.cxx

```diff
@@ -333,148 +378,208 @@ TGeoHMatrix* GeometryTGeo::extractMatrixSensor(Int_t index) const
-  Int_t chipID = 0;
-  Int_t sensorUniqueID;
-  for (Int_t iHalf = 0; iHalf < 2; iHalf++) {
-    HalfSegmentation * halfSeg = seg->getHalf(iHalf);
-    for (Int_t iDisk = 0; iDisk < mNumberOfDisks; iDisk++) {
-      HalfDiskSegmentation* halfDiskSeg = halfSeg->getHalfDisk(iDisk);
-      for (Int_t iLadder = 0; iLadder < halfDiskSeg->getNLadders(); iLadder++) {
-	LadderSegmentation* ladderSeg = halfDiskSeg->getLadder(iLadder);
-	mNumberOfChips += ladderSeg->getNSensors();
-	for (Int_t iSensor = 0; iSensor < ladderSeg->getNSensors(); iSensor++) {
-	  ChipSegmentation * chipSeg = ladderSeg->getSensor(iSensor);
-	  sensorUniqueID = chipSeg->GetUniqueID();
-	  mChipIDmap.insert(BimapType::value_type(sensorUniqueID,chipID++));
-	}
-      }
+  //LOG(INFO) << "mask " << mask << " bit2Mask " << bit2Mask(TransformType::L2G) << FairLogger::endl;
+  // build matrices
+  if ( (mask & bit2Mask(TransformType::L2G)) && !getCacheL2G().isFilled() ) {
+    LOG(INFO) << "Loading ITS L2G matrices from TGeo" <<  FairLogger::endl;
+    auto& cacheL2G = getCacheL2G();
+    cacheL2G.setSize(mSize);
+    for (Int_t i = 0; i < mSize; i++) {
+      TGeoHMatrix* hm = extractMatrixSensor(i);
+      cacheL2G.setMatrix( hm ? Mat3D(*hm) : Mat3D(), i);
     }
-  }
-  
-  mChipHalfID = new Int_t[mNumberOfChips];
-  mChipDiskID = new Int_t[mNumberOfChips];
-  mChipPlaneID = new Int_t[mNumberOfChips];
-  mChipLadderID = new Int_t[mNumberOfChips];
-  mChipSensorID = new Int_t[mNumberOfChips];
-  for (Int_t iHalf = 0; iHalf < 2; iHalf++) {
-    HalfSegmentation * halfSeg = seg->getHalf(iHalf);
-    for (Int_t iDisk = 0; iDisk < mNumberOfDisks; iDisk++) {
-      HalfDiskSegmentation* halfDiskSeg = halfSeg->getHalfDisk(iDisk);
-      for (Int_t iLadder = 0; iLadder < halfDiskSeg->getNLadders(); iLadder++) {
-	LadderSegmentation* ladderSeg = halfDiskSeg->getLadder(iLadder);
-	for (Int_t iSensor = 0; iSensor < ladderSeg->getNSensors(); iSensor++) {
-	  ChipSegmentation * chipSeg = ladderSeg->getSensor(iSensor);
-	  for (Int_t plane = 0; plane < 2; plane++) {
-	    sensorUniqueID = mftGeo->getObjectID(Geometry::SensorType,iHalf,iDisk,plane,iLadder,iSensor);
-	    for (BimapType::left_map::const_iterator left_iter = mChipIDmap.left.begin(), iend = mChipIDmap.left.end(); left_iter != iend; ++left_iter) {
-	      if (left_iter->first == sensorUniqueID) {
-		chipID = left_iter->second; 
-		mChipHalfID[chipID] = iHalf;
-		mChipDiskID[chipID] = iDisk;
-		mChipPlaneID[chipID] = plane;
-		mChipLadderID[chipID] = iLadder;
-		mChipSensorID[chipID] = iSensor;
-	      }
-	    }
-	  }
-	}
-      }
+  } 
+
+  if ( (mask & bit2Mask(TransformType::T2L)) && !getCacheT2L().isFilled() ) {
+    // matrices for Tracking to Local (Sensor!!! rather than the full chip) frame transformation
+    LOG(INFO) << "Loading ITS T2L matrices from TGeo" <<  FairLogger::endl;
+    auto& cacheT2L = getCacheT2L();
+    cacheT2L.setSize(mSize);
+    for (int i=0;i<mSize;i++) {
+      TGeoHMatrix& hm = createT2LMatrix(i);
```

Comment:
If you don't use Tracking2Local matrices, it is better not to fill them with dummy ones and it is done currently (you can even delete the method getMatrixT2L) but I guess you need them for reconstruction?